Cant compil

Command:
bash build.sh


```error[E0308]: mismatched types
  --> src/printer.rs:19:35
   |
19 |         tree_of_cells_into_base64(state.data.as_ref()),
   |                                   ^^^^^^^^^^^^^^^^^^^ expected struct `ton_types::Cell`, found struct `ton_types::cell::Cell`
   |
   = note: expected enum `std::option::Option<&ton_types::Cell>`
              found enum `std::option::Option<&ton_types::cell::Cell>`
   = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
  --> src/printer.rs:20:35
   |
20 |         tree_of_cells_into_base64(state.code.as_ref()),
   |                                   ^^^^^^^^^^^^^^^^^^^ expected struct `ton_types::Cell`, found struct `ton_types::cell::Cell`
   |
   = note: expected enum `std::option::Option<&ton_types::Cell>`
              found enum `std::option::Option<&ton_types::cell::Cell>`
   = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
  --> src/printer.rs:21:35
   |
21 |         tree_of_cells_into_base64(state.library.root()),
   |                                   ^^^^^^^^^^^^^^^^^^^^ expected struct `ton_types::Cell`, found struct `ton_types::cell::Cell`
   |
   = note: expected enum `std::option::Option<&ton_types::Cell>`
              found enum `std::option::Option<&ton_types::cell::Cell>`
   = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
  --> src/printer.rs:52:13
   |
52 | /             msg.body()
53 | |                 .map(|slice| slice.into_cell())
54 | |                 .as_ref(),
   | |_________________________^ expected struct `ton_types::Cell`, found struct `ton_types::cell::Cell`
   |
   = note: expected enum `std::option::Option<&ton_types::Cell>`
              found enum `std::option::Option<&ton_types::cell::Cell>`
   = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
  --> src/util.rs:56:44
   |
56 |     MsgAddressInt::with_standart(None, wc, AccountId::from(address)).unwrap()
   |                                            ^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `ton_types::cell::slice::SliceData`, found struct `SliceData`
   |
   = note: perhaps two different versions of crate `ton_types` are being used?

error[E0277]: the trait bound `ton_types::cell::slice::SliceData: From<ton_types::Cell>` is not satisfied
  --> src/util.rs:65:31
   |
65 |     StateInit::construct_from(&mut cell.into()).unwrap()
   |                               ^^^^^^^^^^^^^^^^ the trait `From<ton_types::Cell>` is not implemented for `ton_types::cell::slice::SliceData`
   |
   = help: the following implementations were found:
             <ton_types::cell::slice::SliceData as From<&[u8]>>
             <ton_types::cell::slice::SliceData as From<&mut ton_types::cell::builder::BuilderData>>
             <ton_types::cell::slice::SliceData as From<&ton_types::cell::Cell>>
             <ton_types::cell::slice::SliceData as From<&ton_types::cell::builder::BuilderData>>
           and 6 others
   = note: required because of the requirements on the impl of `Into<ton_types::cell::slice::SliceData>` for `ton_types::Cell`

error[E0308]: mismatched types
  --> src/util.rs:74:23
   |
74 |     *msg.body_mut() = body;
   |                       ^^^^ expected struct `ton_types::cell::slice::SliceData`, found struct `SliceData`
   |
   = note: expected enum `std::option::Option<ton_types::cell::slice::SliceData>`
              found enum `std::option::Option<SliceData>`
   = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
  --> src/util.rs:99:23
   |
99 |     *msg.body_mut() = body;
   |                       ^^^^ expected struct `ton_types::cell::slice::SliceData`, found struct `SliceData`
   |
   = note: expected enum `std::option::Option<ton_types::cell::slice::SliceData>`
              found enum `std::option::Option<SliceData>`
   = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
  --> src/abi.rs:57:61
   |
57 |             let res = decode_unknown_function_call(abi_str, body.clone(), internal);
   |                                                             ^^^^^^^^^^^^ expected struct `ton_types::cell::slice::SliceData`, found struct `SliceData`
   |
   = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
   --> src/abi.rs:132:57
    |
132 |     if let Some(res) = gs.all_abis.decode_function_call(&body, internal) {
    |                                                         ^^^^^ expected struct `SliceData`, found struct `ton_types::cell::slice::SliceData`
    |
    = note: expected reference `&SliceData`
               found reference `&ton_types::cell::slice::SliceData`
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
   --> src/abi.rs:154:5
    |
153 |   ) -> Result<BuilderData, String> {
    |        --------------------------- expected `Result<BuilderData, std::string::String>` because of return type
154 | /     encode_function_call(
155 | |         abi_info.text().clone(),
156 | |         method.to_owned(),
157 | |         header,
...   |
160 | |         pair,
161 | |     ).map_err(|e| format!("cannot encode abi body: {:?}", e))
    | |_____________________________________________________________^ expected struct `BuilderData`, found struct `ton_types::cell::builder::BuilderData`
    |
    = note: expected enum `Result<BuilderData, _>`
               found enum `Result<ton_types::cell::builder::BuilderData, _>`
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
  --> src/actions.rs:94:33
   |
94 |             state_init.set_code(c);
   |                                 ^ expected struct `ton_types::cell::Cell`, found struct `ton_types::Cell`
   |
   = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
   --> src/actions.rs:179:31
    |
179 |             state.code = Some(new_code);
    |                               ^^^^^^^^ expected struct `ton_types::Cell`, found struct `ton_types::cell::Cell`
    |
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
   --> src/debug_info.rs:109:49
    |
109 |     let dict1 = HashmapE::with_hashmap(32, Some(root_cell.reference(0).unwrap()));
    |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `ton_types::Cell`, found struct `ton_types::cell::Cell`
    |
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
   --> src/debug_info.rs:110:49
    |
110 |     let dict2 = HashmapE::with_hashmap(32, Some(root_cell.reference(1).unwrap().reference(0).unwrap()));
    |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `ton_types::Cell`, found struct `ton_types::cell::Cell`
    |
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0277]: the trait bound `SliceData: From<ton_types::cell::Cell>` is not satisfied
   --> src/debug_info.rs:114:72
    |
114 |         let key = id.write_to_new_cell().unwrap().into_cell().unwrap().into();
    |                                                                        ^^^^ the trait `From<ton_types::cell::Cell>` is not implemented for `SliceData`
    |
    = help: the following implementations were found:
              <SliceData as From<&BuilderData>>
              <SliceData as From<&UInt256>>
              <SliceData as From<&[u8]>>
              <SliceData as From<&mut BuilderData>>
            and 6 others
    = note: required because of the requirements on the impl of `Into<SliceData>` for `ton_types::cell::Cell`

error[E0277]: the trait bound `SliceData: From<ton_types::cell::Cell>` is not satisfied
   --> src/debug_info.rs:134:72
    |
134 |         let key = id.write_to_new_cell().unwrap().into_cell().unwrap().into();
    |                                                                        ^^^^ the trait `From<ton_types::cell::Cell>` is not implemented for `SliceData`
    |
    = help: the following implementations were found:
              <SliceData as From<&BuilderData>>
              <SliceData as From<&UInt256>>
              <SliceData as From<&[u8]>>
              <SliceData as From<&mut BuilderData>>
            and 6 others
    = note: required because of the requirements on the impl of `Into<SliceData>` for `ton_types::cell::Cell`

error[E0277]: the trait bound `SliceData: From<ton_types::cell::Cell>` is not satisfied
   --> src/debug_info.rs:154:72
    |
154 |         let key = id.write_to_new_cell().unwrap().into_cell().unwrap().into();
    |                                                                        ^^^^ the trait `From<ton_types::cell::Cell>` is not implemented for `SliceData`
    |
    = help: the following implementations were found:
              <SliceData as From<&BuilderData>>
              <SliceData as From<&UInt256>>
              <SliceData as From<&[u8]>>
              <SliceData as From<&mut BuilderData>>
            and 6 others
    = note: required because of the requirements on the impl of `Into<SliceData>` for `ton_types::cell::Cell`

error[E0308]: mismatched types
   --> src/debug_info.rs:160:26
    |
160 |     hash2function.insert(root_cell.repr_hash(), "selector".to_owned());
    |                          ^^^^^^^^^^^^^^^^^^^^^ expected struct `UInt256`, found struct `ton_types::types::UInt256`
    |
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
   --> src/debug_info.rs:162:30
    |
162 |         hash2function.insert(selector2.repr_hash(), "selector2".to_owned());
    |                              ^^^^^^^^^^^^^^^^^^^^^ expected struct `UInt256`, found struct `ton_types::types::UInt256`
    |
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
   --> src/exec.rs:102:36
    |
102 |     let address0 = convert_address(state_init.hash().unwrap(), wc);
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `UInt256`, found struct `ton_types::types::UInt256`
    |
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0277]: the trait bound `ton_types::Cell: From<ton_types::cell::Cell>` is not satisfied
   --> src/call_contract.rs:131:40
    |
131 |         let msg_cell = StackItem::Cell(msg.serialize().unwrap().into());
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<ton_types::cell::Cell>` is not implemented for `ton_types::Cell`
    |
    = help: the following implementations were found:
              <ton_types::Cell as From<&BuilderData>>
              <ton_types::Cell as From<&mut BuilderData>>
              <ton_types::Cell as From<BuilderData>>
    = note: required because of the requirements on the impl of `Into<ton_types::Cell>` for `ton_types::cell::Cell`

error[E0277]: the trait bound `SliceData: From<ton_types::cell::slice::SliceData>` is not satisfied
   --> src/call_contract.rs:134:26
    |
134 |             Some(b) => b.into(),
    |                          ^^^^ the trait `From<ton_types::cell::slice::SliceData>` is not implemented for `SliceData`
    |
    = help: the following implementations were found:
              <SliceData as From<&BuilderData>>
              <SliceData as From<&UInt256>>
              <SliceData as From<&[u8]>>
              <SliceData as From<&mut BuilderData>>
            and 6 others
    = note: required because of the requirements on the impl of `Into<SliceData>` for `ton_types::cell::slice::SliceData`

error[E0308]: mismatched types
   --> src/call_contract.rs:212:40
    |
212 |                 state_init.data = Some(root_cell);
    |                                        ^^^^^^^^^ expected struct `ton_types::cell::Cell`, found struct `ton_types::Cell`
    |
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0277]: the trait bound `ton_types::cell::slice::SliceData: From<ton_types::Cell>` is not satisfied
   --> src/call_contract.rs:229:40
    |
229 |             OutActions::construct_from(&mut cell.into()).unwrap(),
    |                                        ^^^^^^^^^^^^^^^^ the trait `From<ton_types::Cell>` is not implemented for `ton_types::cell::slice::SliceData`
    |
    = help: the following implementations were found:
              <ton_types::cell::slice::SliceData as From<&[u8]>>
              <ton_types::cell::slice::SliceData as From<&mut ton_types::cell::builder::BuilderData>>
              <ton_types::cell::slice::SliceData as From<&ton_types::cell::Cell>>
              <ton_types::cell::slice::SliceData as From<&ton_types::cell::builder::BuilderData>>
            and 6 others
    = note: required because of the requirements on the impl of `Into<ton_types::cell::slice::SliceData>` for `ton_types::Cell`

error[E0277]: the trait bound `ton_types::cell::Cell: From<BuilderData>` is not satisfied
   --> src/call_contract.rs:259:43
    |
259 |             .unwrap_or(BuilderData::new().into())
    |                                           ^^^^ the trait `From<BuilderData>` is not implemented for `ton_types::cell::Cell`
    |
    = help: the following implementations were found:
              <ton_types::cell::Cell as From<&mut ton_types::cell::builder::BuilderData>>
              <ton_types::cell::Cell as From<&ton_types::cell::builder::BuilderData>>
              <ton_types::cell::Cell as From<ton_types::cell::builder::BuilderData>>
    = note: required because of the requirements on the impl of `Into<ton_types::cell::Cell>` for `BuilderData`

error[E0277]: the trait bound `SliceData: From<ton_types::cell::Cell>` is not satisfied
   --> src/call_contract.rs:260:14
    |
260 |             .into();
    |              ^^^^ the trait `From<ton_types::cell::Cell>` is not implemented for `SliceData`
    |
    = help: the following implementations were found:
              <SliceData as From<&BuilderData>>
              <SliceData as From<&UInt256>>
              <SliceData as From<&[u8]>>
              <SliceData as From<&mut BuilderData>>
            and 6 others
    = note: required because of the requirements on the impl of `Into<SliceData>` for `ton_types::cell::Cell`

error[E0277]: the trait bound `ton_types::cell::Cell: From<BuilderData>` is not satisfied
   --> src/call_contract.rs:263:43
    |
263 |             .unwrap_or(BuilderData::new().into())
    |                                           ^^^^ the trait `From<BuilderData>` is not implemented for `ton_types::cell::Cell`
    |
    = help: the following implementations were found:
              <ton_types::cell::Cell as From<&mut ton_types::cell::builder::BuilderData>>
              <ton_types::cell::Cell as From<&ton_types::cell::builder::BuilderData>>
              <ton_types::cell::Cell as From<ton_types::cell::builder::BuilderData>>
    = note: required because of the requirements on the impl of `Into<ton_types::cell::Cell>` for `BuilderData`

error[E0277]: the trait bound `SliceData: From<ton_types::cell::Cell>` is not satisfied
   --> src/call_contract.rs:264:14
    |
264 |             .into();
    |              ^^^^ the trait `From<ton_types::cell::Cell>` is not implemented for `SliceData`
    |
    = help: the following implementations were found:
              <SliceData as From<&BuilderData>>
              <SliceData as From<&UInt256>>
              <SliceData as From<&[u8]>>
              <SliceData as From<&mut BuilderData>>
            and 6 others
    = note: required because of the requirements on the impl of `Into<SliceData>` for `ton_types::cell::Cell`

error[E0277]: the trait bound `SliceData: From<ton_types::cell::Cell>` is not satisfied
   --> src/call_contract.rs:277:51
    |
277 |     let mut info = SmartContractInfo::with_myself(myself.serialize().unwrap().into());
    |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<ton_types::cell::Cell>` is not implemented for `SliceData`
    |
    = help: the following implementations were found:
              <SliceData as From<&BuilderData>>
              <SliceData as From<&UInt256>>
              <SliceData as From<&[u8]>>
              <SliceData as From<&mut BuilderData>>
            and 6 others
    = note: required because of the requirements on the impl of `Into<SliceData>` for `ton_types::cell::Cell`

error[E0308]: mismatched types
   --> src/call_contract.rs:279:43
    |
279 |     *info.balance_remaining_other_mut() = balance.1.other_as_hashmap().clone();
    |                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `HashmapE`, found struct `ton_types::dictionary::hashmap::HashmapE`
    |
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
   --> src/messages.rs:475:29
    |
475 |         b.append_bytestring(&body).unwrap();
    |                             ^^^^^ expected struct `SliceData`, found struct `ton_types::cell::slice::SliceData`
    |
    = note: expected reference `&SliceData`
               found reference `&ton_types::cell::slice::SliceData`
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0277]: the trait bound `ton_types::cell::slice::SliceData: From<BuilderData>` is not satisfied
   --> src/messages.rs:531:25
    |
531 |                         BuilderData::with_raw(vec![0x55; 8], 64).unwrap().into()
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<BuilderData>` is not implemented for `ton_types::cell::slice::SliceData`
    |
    = help: the following implementations were found:
              <ton_types::cell::slice::SliceData as From<&[u8]>>
              <ton_types::cell::slice::SliceData as From<&mut ton_types::cell::builder::BuilderData>>
              <ton_types::cell::slice::SliceData as From<&ton_types::cell::Cell>>
              <ton_types::cell::slice::SliceData as From<&ton_types::cell::builder::BuilderData>>
            and 6 others
    = note: required because of the requirements on the impl of `Into<ton_types::cell::slice::SliceData>` for `BuilderData`

error[E0308]: mismatched types
   --> src/lib.rs:156:30
    |
156 |     let code = serialize_toc(&code).unwrap();
    |                              ^^^^^ expected struct `ton_types::Cell`, found `&ton_types::cell::Cell`
    |
    = note: expected reference `&ton_types::Cell`
               found reference `&&ton_types::cell::Cell`

error[E0308]: mismatched types
   --> src/lib.rs:159:30
    |
159 |     let data = serialize_toc(&data).unwrap();
    |                              ^^^^^ expected struct `ton_types::Cell`, found `&ton_types::cell::Cell`
    |
    = note: expected reference `&ton_types::Cell`
               found reference `&&ton_types::cell::Cell`

error[E0277]: the trait bound `ton_types::Cell: From<ton_types::cell::builder::BuilderData>` is not satisfied
   --> src/lib.rs:173:10
    |
173 |         .into();
    |          ^^^^ the trait `From<ton_types::cell::builder::BuilderData>` is not implemented for `ton_types::Cell`
    |
    = help: the following implementations were found:
              <ton_types::Cell as From<&BuilderData>>
              <ton_types::Cell as From<&mut BuilderData>>
              <ton_types::Cell as From<BuilderData>>
    = note: required because of the requirements on the impl of `Into<ton_types::Cell>` for `ton_types::cell::builder::BuilderData`

error[E0308]: mismatched types
   --> src/lib.rs:393:31
    |
393 |     let bytes = serialize_toc(&code).unwrap();
    |                               ^^^^^ expected struct `ton_types::Cell`, found struct `ton_types::cell::Cell`
    |
    = note: expected reference `&ton_types::Cell`
               found reference `&ton_types::cell::Cell`
    = note: perhaps two different versions of crate `ton_types` are being used?

error[E0308]: mismatched types
   --> src/lib.rs:402:31
    |
402 |     let bytes = serialize_toc(&data).unwrap();
    |                               ^^^^^ expected struct `ton_types::Cell`, found struct `ton_types::cell::Cell`
    |
    = note: expected reference `&ton_types::Cell`
               found reference `&ton_types::cell::Cell`
    = note: perhaps two different versions of crate `ton_types` are being used?

```

--- a/linker_lib/Cargo.toml
+++ b/linker_lib/Cargo.toml
@@ -19,7 +19,8 @@ rand = "0.7.3"
 serde = { version = "1.0.125", features = ["derive"]  }
 serde_json = "1.0.64"
 
-ton_abi = { git = "https://github.com/tonlabs/ton-labs-abi.git" }
-ton_block = { git = "https://github.com/tonlabs/ton-labs-block.git" }
-ton_types = { git = "https://github.com/tonlabs/ton-labs-types.git" }
-ton_vm = { git = "https://github.com/tonlabs/ton-labs-vm.git", default-features = false }
+ton_abi = { git = "https://github.com/tonlabs/ton-labs-abi.git", tag = "2.1.3" }
+ton_block = { git = "https://github.com/tonlabs/ton-labs-block.git", tag = "1.7.27" }
+ton_types = { git = "https://github.com/tonlabs/ton-labs-types.git", tag = "1.10.10" }
+ton_vm = { git = "https://github.com/tonlabs/ton-labs-vm.git", tag = "1.8.19", default-features = false }
+

